File: app/src/main/java/com/moronlu18/savestateactivity/AnswerAdapter.java
package com.moronlu18.savestateactivity;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.RadioButton;
import android.widget.TextView;


public class AnswerAdapter extends ArrayAdapter<Answer> {
    private Context context;
    int answerPosition = 0;

    public AnswerAdapter(Context context) {
        super(context, R.layout.item_task);
        this.context=context;
    }


    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {

        final AnswerHolder answerHolder;
        if (convertView == null) {
            LayoutInflater vi = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = vi.inflate(R.layout.item_task, null);
            answerHolder=new AnswerHolder();
            answerHolder.rdAnswer=  convertView.findViewById(R.id.rdAnswer);
            answerHolder.txvAnswer=convertView.findViewById(R.id.txvAnswer);
            convertView.setTag(answerHolder);

        }
        else{
            answerHolder=(AnswerHolder)convertView.getTag();
        }
        answerHolder.txvAnswer.setText(getItem(position).getText());
        answerHolder.rdAnswer.setChecked(position == answerPosition);

        answerHolder.rdAnswer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                answerPosition = position;
                getItem(position).setSelection(!getItem(position).isSelection());
                notifyDataSetChanged();
            }
        });
        return convertView;
    }

    public void setAnswer() {
        getItem(answerPosition).setSelection(!getItem(answerPosition).isSelection());
        notifyDataSetChanged();
    }

    class AnswerHolder{
        RadioButton rdAnswer;
        TextView txvAnswer;

    }
}

File: app/src/main/java/com/example/savestateactivity/AnswerAdapter.kt
package com.example.savestateactivity

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.savestateactivity.databinding.ItemTaskBinding

class AnswerAdapter() :
    RecyclerView.Adapter<AnswerAdapter.AnswerViewHolder>() {

    private var answers: ArrayList<Answer> = ArrayList()

    init {
        answers.add(Answer(true, "Opcion A"))
        answers.add(Answer(false, "Opcion B"))
        answers.add(Answer(false, "Opcion C"))
        answers.add(Answer(false, "Opción D"))
        answers.add(Answer(false, "Opción E"))
        answers.add(Answer(false, "Opción F"))
    }

    var answerPosition = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AnswerViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ItemTaskBinding.inflate(inflater, parent, false)
        return AnswerViewHolder(binding)
    }

    override fun onBindViewHolder(holder: AnswerViewHolder, position: Int) {
        val answer = answers[position]
        holder.bind(answer)
    }

    override fun getItemCount(): Int {
        return answers.size
    }

    inner class AnswerViewHolder(private val binding: ItemTaskBinding) :
        RecyclerView.ViewHolder(binding.root) {

        init {
            binding.rdAnswer.setOnClickListener {
                val currentPosition = adapterPosition
                if (currentPosition != RecyclerView.NO_POSITION) {
                    answerPosition = currentPosition
                    answers[currentPosition].isSelection = !answers[currentPosition].isSelection
                    notifyDataSetChanged()
                }
            }
        }

        fun bind(answer: Answer) {
            binding.txvAnswer.text = answer.text
            binding.rdAnswer.isChecked = adapterPosition == answerPosition
        }
    }
}
